### PR TITLE
HBASE-22577 BufferedMutatorOverAsyncBufferedMutator.tryCompleteFuture…

### PR DESCRIPTION
**Pull request by Apache9**
… consume too much CPU time

### CODE REVIEW

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java [hunk 1]
```diff
@@ -100,62 +103,63 @@ private RetriesExhaustedWithDetailsException makeError() {
...
 
+  private void internalFlush() throws RetriesExhaustedWithDetailsException {
+    // should get the future array before calling mutator.flush, otherwise we may hit an infinite
+    // wait, since someone may add new future to the map after we calling the flush.
+    CompletableFuture<?>[] toWait = futures.keySet().toArray(new CompletableFuture<?>[0]);
+    mutator.flush();
+    try {
+      CompletableFuture.allOf(toWait).join();
+    } catch (CompletionException e) {
+      // just ignore, we will record the actual error in the errors field
```
Comment (busbey) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java]: please add a DEBUG or TRACE log message with the exception. If something goes wrong with the errors field handling we need some way for operators to get a log message with what's going on.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 155 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| 0 | mvndep | 35 | Maven dependency ordering for branch |
| +1 | mvninstall | 268 | HBASE-21512 passed |
| +1 | compile | 77 | HBASE-21512 passed |
| +1 | checkstyle | 103 | HBASE-21512 passed |
| +1 | shadedjars | 275 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 236 | HBASE-21512 passed |
| +1 | javadoc | 54 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 14 | Maven dependency ordering for patch |
| +1 | mvninstall | 250 | the patch passed |
| +1 | compile | 77 | the patch passed |
| +1 | javac | 77 | the patch passed |
| +1 | checkstyle | 104 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 278 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 772 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 279 | the patch passed |
| +1 | javadoc | 53 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 100 | hbase-client in the patch passed. |
| +1 | unit | 11908 | hbase-server in the patch passed. |
| +1 | asflicense | 53 | The patch does not generate ASF License warnings. |
| | | 15436 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/305 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux f122719bee34 4.4.0-137-generic #163~14.04.1-Ubuntu SMP Mon Sep 24 17:14:57 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / 5ca23c3c20 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/2/testReport/ |
| Max. process+thread count | 5028 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 206 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| 0 | mvndep | 34 | Maven dependency ordering for branch |
| +1 | mvninstall | 273 | HBASE-21512 passed |
| +1 | compile | 76 | HBASE-21512 passed |
| +1 | checkstyle | 94 | HBASE-21512 passed |
| +1 | shadedjars | 258 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 254 | HBASE-21512 passed |
| +1 | javadoc | 53 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 15 | Maven dependency ordering for patch |
| +1 | mvninstall | 256 | the patch passed |
| +1 | compile | 77 | the patch passed |
| +1 | javac | 77 | the patch passed |
| -1 | checkstyle | 28 | hbase-client: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 260 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 803 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 280 | the patch passed |
| +1 | javadoc | 53 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 106 | hbase-client in the patch passed. |
| -1 | unit | 15056 | hbase-server in the patch failed. |
| +1 | asflicense | 50 | The patch does not generate ASF License warnings. |
| | | 18652 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.client.TestFromClientSide |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/305 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux f065a0be9d94 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / 93a6aaa545 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/3/artifact/out/diff-checkstyle-hbase-client.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/3/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/3/testReport/ |
| Max. process+thread count | 4723 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/3/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
Any other concerns? @busbey

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java [hunk 2]
```diff
@@ -25,32 +25,39 @@
...
 class BufferedMutatorOverAsyncBufferedMutator implements BufferedMutator {
 
+  private static final Logger LOG = LoggerFactory.getLogger(BufferedMutatorOverAsyncBufferedMutator.class);
+
   private final AsyncBufferedMutator mutator;
 
   private final ExceptionListener listener;
 
-  private List<CompletableFuture<Void>> futures = new ArrayList<>();
+  private final ConcurrentMap<CompletableFuture<Void>, CompletableFuture<Void>> futures =
```
Comment (busbey) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java]: we're just using this as a set of futures right? Would be clearer if we used [ConcurrentHashMap.keySet](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#newKeySet--) to make that explicit.

e.g.

```
private final Set<CompletableFuture<Void>> futures = ConcurrentHashMap.newKeySet()
```

File: hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java [hunk 3]
```diff
@@ -100,62 +107,64 @@ private RetriesExhaustedWithDetailsException makeError() {
...
       });
-      toBuffered.add(toComplete);
     }
     synchronized (this) {
-      futures.addAll(toBuffered);
-      if (futures.size() > BUFFERED_FUTURES_THRESHOLD) {
-        tryCompleteFuture();
-      }
-      if (!errors.isEmpty()) {
+      if (bufferedSize.get() > mutator.getWriteBufferSize() * 2) {
```
Comment (openinx) [inline hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java]: Why bufferSize * 2 here ? maybe need a comment ?

Comment (Apache9) [reply hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorOverAsyncBufferedMutator.java]: The comment is just below this line. It is not a good idea to just check for writeBufferSize as we will likely to always enter the branch, so a * 2 is used.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 143 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for branch |
| +1 | mvninstall | 251 | HBASE-21512 passed |
| +1 | compile | 74 | HBASE-21512 passed |
| +1 | checkstyle | 97 | HBASE-21512 passed |
| +1 | shadedjars | 257 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 234 | HBASE-21512 passed |
| +1 | javadoc | 55 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 17 | Maven dependency ordering for patch |
| +1 | mvninstall | 255 | the patch passed |
| +1 | compile | 79 | the patch passed |
| +1 | javac | 79 | the patch passed |
| -1 | checkstyle | 29 | hbase-client: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 260 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 727 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 267 | the patch passed |
| +1 | javadoc | 57 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 108 | hbase-client in the patch passed. |
| -1 | unit | 15277 | hbase-server in the patch failed. |
| +1 | asflicense | 55 | The patch does not generate ASF License warnings. |
| | | 18687 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestFromClientSide |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.quotas.TestQuotaAdmin |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/305 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux b4d16958a66c 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / c3420640f5 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/4/artifact/out/diff-checkstyle-hbase-client.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/4/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/4/testReport/ |
| Max. process+thread count | 5289 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/4/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache9** [comment]
Any other concerns? Thanks.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 45 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ HBASE-21512 Compile Tests _ |
| 0 | mvndep | 29 | Maven dependency ordering for branch |
| +1 | mvninstall | 254 | HBASE-21512 passed |
| +1 | compile | 77 | HBASE-21512 passed |
| +1 | checkstyle | 94 | HBASE-21512 passed |
| +1 | shadedjars | 263 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 255 | HBASE-21512 passed |
| +1 | javadoc | 55 | HBASE-21512 passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 16 | Maven dependency ordering for patch |
| +1 | mvninstall | 249 | the patch passed |
| +1 | compile | 79 | the patch passed |
| +1 | javac | 79 | the patch passed |
| +1 | checkstyle | 94 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 278 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 904 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 318 | the patch passed |
| +1 | javadoc | 54 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 244 | hbase-client in the patch passed. |
| -1 | unit | 17050 | hbase-server in the patch failed. |
| +1 | asflicense | 44 | The patch does not generate ASF License warnings. |
| | | 20742 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestFromClientSide |
|   | hadoop.hbase.client.TestSnapshotTemporaryDirectoryWithRegionReplicas |
|   | hadoop.hbase.client.TestCloneSnapshotFromClientNormal |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/305 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 9dfd0754527d 4.4.0-137-generic #163-Ubuntu SMP Mon Sep 24 13:14:43 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21512 / 948f33bc4c |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/5/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/5/testReport/ |
| Max. process+thread count | 4860 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-305/5/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Pull request merged**